File: src/main/java/net/ofk/dbmapper/defaults/impl/DefaultStorage.java
package net.ofk.dbmapper.defaults.impl;

import net.ofk.dbmapper.defaults.api.Engine;
import net.ofk.dbmapper.api.Storage;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.lang.reflect.Constructor;
import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.ResultSetMetaData;
import java.sql.Statement;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

/**
 * An implementation of the storage which is needed only and created by the default transaction implementation.
 */
class DefaultStorage implements Storage<ResultSet> {
  private static final Logger LOG = LoggerFactory.getLogger(DefaultStorage.class);

  private final Connection conn;
  private final Engine engine;

  DefaultStorage(final Connection conn, final Engine engine) {
    this.conn = conn;
    this.engine = engine;
  }

  @Override
  public int update(final String queryTemplate, final Object... paramValues) throws Exception {
    String query = this.engine.buildQuery(queryTemplate, paramValues);

    Integer result = null;
    try (
      Statement st = this.conn.createStatement();
    ) {
      try {
        result = st.executeUpdate(query);
      } catch (final Exception ex) {
        throw this.handleStatementError(ex, query);
      }
    }

    return result;
  }

  @Override
  public List<?> insert(final String queryTemplate, final Object... paramValues) throws Exception {
    List<Object> result = new ArrayList<>();

    String query = this.engine.buildQuery(queryTemplate, paramValues);

    try (
      Statement st = this.conn.createStatement();
    ) {
      try {
        st.execute(query, Statement.RETURN_GENERATED_KEYS);
      } catch (final Exception ex) {
        throw this.handleStatementError(ex, query);
      }

      try (
        ResultSet rs = st.getGeneratedKeys();
      ) {
        while (rs.next()) {
          result.add(rs.getObject(1));
        }
      }
    }

    return result;
  }

  @Override
  public <T> List<T> select(final Mapper<ResultSet, T> mapper, final String queryTemplate, final Object... paramValues) throws Throwable {
    return this.doSelect(rs -> this.map(rs, mapper), queryTemplate, paramValues);
  }

  @Override
  public <T> List<T> select(final Class<T> type, final String queryTemplate, final Object... paramValues) throws Throwable {
    return this.doSelect(rs -> this.mapToClass(rs, type), queryTemplate, paramValues);
  }

  @Override
  public List<List<?>> select(final String queryTemplate, final Object... paramValues) throws Throwable {
    return this.doSelect(this::mapToList, queryTemplate, paramValues);
  }

  private <T> List<T> doSelect(
    final Mapper<ResultSet, List<T>> mapper,
    final String queryTemplate,
    final Object... paramValues
  ) throws Throwable {
    List<T> result = null;

    String query = this.engine.buildQuery(queryTemplate, paramValues);

    try (
      Statement st = this.conn.createStatement();
    ) {
      ResultSet rs = null;
      try {
        try {
          rs = st.executeQuery(query);
        } catch (final Throwable ex) {
          throw this.handleStatementError(ex, query);
        }

        result = mapper.map(rs);
      } finally {
        if (rs != null) {
          rs.close();
        }
      }
    }

    return result;
  }

  private <T> List<T> mapToClass(final ResultSet rs, final Class<T> type) throws Exception {
    List<T> result = new ArrayList<>();

    ResultSetMetaData metaData = rs.getMetaData();
    Class<?>[] columnClasses = new Class[metaData.getColumnCount()];
    for (int i = 1; i <= metaData.getColumnCount(); i++) {
      columnClasses[i - 1] = Class.forName(metaData.getColumnClassName(i));
    }

    Object[] columnValues = new Object[metaData.getColumnCount()];
    while (rs.next()) {
      for (int i = 1; i <= metaData.getColumnCount(); i++) {
        columnValues[i - 1] = rs.getObject(i);
      }

      Constructor<T> ctr = type.getConstructor(columnClasses);
      result.add(ctr.newInstance(columnValues));
    }

    return result;
  }

  private List<List<?>> mapToList(final ResultSet rs) throws Exception {
    List<List<?>> result = new ArrayList<>();
    ResultSetMetaData metaData = rs.getMetaData();
    while (rs.next()) {
      List<Object> values = new ArrayList<>();
      for (int i = 1; i <= metaData.getColumnCount(); i++) {
        values.add(rs.getObject(i));
      }
      result.add(values);
    }
    return result;
  }

  private <T> List<T> map(final ResultSet rs, final Mapper<ResultSet, T> mapper) throws Throwable {
    List<T> result = new ArrayList<>();
    while (rs.next()) {
      result.add(mapper.map(rs));
    }
    return result;
  }

  @Override
  public void updateMany(final int count) {
    if (count <= 0) { throw new IllegalStateException(); }
  }

  @Override
  public void updateOne(final int count) {
    if (count != 1) { throw new IllegalStateException(); }
  }

  @Override
  public void updateOneOrNone(final int count) {
    if (count != 0 && count != 1) { throw new IllegalStateException(); }
  }

  @Override
  public <T> List<T> takeMany(final List<T> list) {
    this.updateMany(list.size());
    return list;
  }

  @Override
  public <T> T takeOne(final List<T> list) {
    this.updateOne(list.size());
    return list.iterator().next();
  }

  @Override
  public <T> T takeOneOrNone(final List<T> list) {
    this.updateOneOrNone(list.size());
    return list.isEmpty() ? null : list.iterator().next();
  }

  @Override
  public <T> List<T> takeFirstColumn(final List<List<?>> list) {
    return list.stream().map(e -> (T) e.iterator().next()).collect(Collectors.toList());
  }

  private <E> E handleStatementError(final E ex, final String query) {
    DefaultStorage.LOG.error("Failed query: " + query, ex);
    return ex;
  }

  @Override
  public String escape(final String value) {
    return this.engine.escape(value);
  }
}

File: src/main/kotlin/net/ofk/dbmapper/defaults/impl/DefaultStorage.kt
package net.ofk.dbmapper.defaults.impl

import net.ofk.dbmapper.api.Mapper
import net.ofk.dbmapper.api.Storage
import net.ofk.dbmapper.defaults.api.Engine
import net.ofk.kutils.Auto
import org.slf4j.LoggerFactory
import java.sql.Connection
import java.sql.ResultSet
import java.sql.Statement

/**
 * An implementation of the storage which is needed only and created by the default transaction implementation.
 */
internal class DefaultStorage(
  private val conn: Connection,
  private val engine: Engine
): Storage<ResultSet> {
  companion object {
     private val LOG = LoggerFactory.getLogger(DefaultStorage::class.java)
  }

  override fun update(queryTemplate: String, vararg paramValues: Any?): Int {
    val query = engine.buildQuery(queryTemplate, *paramValues)

    return Auto.close {
      val st = conn.createStatement().open()
      try {
        st!!.executeUpdate(query)
      } catch (ex: Exception) {
        throw handleStatementError(ex, query)
      }
    }
  }

  override fun insert(queryTemplate: String, vararg paramValues: Any?): List<*> {
    val query = engine.buildQuery(queryTemplate, *paramValues)

    return Auto.close {
      val st = conn.createStatement().open()
      try {
        st!!.execute(query, Statement.RETURN_GENERATED_KEYS)
      } catch (ex: Exception) {
        throw handleStatementError(ex, query)
      }

      val result = arrayListOf<Any>()
      val rs = st.generatedKeys.open()
      while (rs!!.next()) {
        result.add(rs.getObject(1))
      }
      result
    }
  }

  override fun <T> select(mapper: (ResultSet) -> T, queryTemplate: String, vararg paramValues: Any?): List<T> {
    return doSelect({rs -> map(rs, mapper)}, queryTemplate, *paramValues)
  }

  override fun <T> sel(mapper: Mapper<ResultSet, T>, queryTemplate: String, vararg paramValues: Any?): List<T> {
    return select({rs -> mapper.map(rs)}, queryTemplate, *paramValues)
  }

  override fun <T> select(type: Class<T>, queryTemplate: String, vararg paramValues: Any?): List<T> {
    return doSelect({rs -> mapToClass(rs, type)}, queryTemplate, *paramValues)
  }

  override fun select(queryTemplate: String, vararg paramValues: Any?): List<List<Any?>> {
    return doSelect({rs -> mapToList(rs)}, queryTemplate, *paramValues)
  }

  private fun <T> doSelect(
    mapper: (ResultSet) -> List<T>,
    queryTemplate: String,
    vararg paramValues: Any?
  ): List<T> {
    val query = engine.buildQuery(queryTemplate, *paramValues)

    return Auto.close {
      val st = conn.createStatement().open()

      val rs = try {
        st!!.executeQuery(query).open()
      } catch (ex: Exception) {
        throw handleStatementError(ex, query)
      }

      mapper.invoke(rs!!)
    }
  }

  private fun <T> mapToClass(rs: ResultSet, type: Class<T>): List<T> {
    val result = arrayListOf<T>();

    val metaData = rs.metaData
    val columnClasses = arrayOfNulls<Class<*>>(metaData.columnCount)
    for (i in 1..metaData.columnCount) {
      columnClasses[i - 1] = Class.forName(metaData.getColumnClassName(i))
    }

    val columnValues = arrayOfNulls<Any?>(metaData.columnCount)
    while (rs.next()) {
      for (i in 1..metaData.columnCount) {
        columnValues[i - 1] = rs.getObject(i)
      }

      val ctr = type.getConstructor(*columnClasses)
      result.add(ctr.newInstance(*columnValues))
    }

    return result
  }

  private fun mapToList(rs: ResultSet): List<List<Any?>> {
    val result = arrayListOf<List<Any?>>()
    val metaData = rs.metaData
    while (rs.next()) {
      val values = arrayListOf<Any?>()
      for (i in 1..metaData.columnCount) {
        values.add(rs.getObject(i))
      }
      result.add(values)
    }
    return result
  }

  private fun <T> map(rs: ResultSet, mapper: (ResultSet) -> T): List<T> {
    val result = arrayListOf<T>()
    while (rs.next()) {
      result.add(mapper.invoke(rs))
    }
    return result
  }

  override fun updateMany(count: Int) {
    if (count <= 0) { throw IllegalStateException() }
  }

  override fun updateOne(count: Int) {
    if (count != 1) { throw IllegalStateException() }
  }

  override fun updateOneOrNone(count: Int) {
    if (count != 0 && count != 1) { throw IllegalStateException() }
  }

  override fun <T> takeMany(list: List<T>): List<T> {
    updateMany(list.size)
    return list
  }

  override fun <T> takeOne(list: List<T>): T {
    updateOne(list.size)
    return list.iterator().next()
  }

  override fun <T> takeOneOrNone(list: List<T>): T? {
    updateOneOrNone(list.size)
    return if (list.isEmpty()) null else list.iterator().next()
  }

  override fun <T> takeFirstColumn(list: List<List<*>>): List<T> = list.map { e -> e.iterator().next() as T }

  private fun <E> handleStatementError(ex: E, query: String): E {
    LOG.error("Failed query: $query", ex)
    return ex
  }
}
